crates/integrations/datafusion/src/physical_plan/scan.rs

```diff
@@ -0,0 +1,123 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::{any::Any, pin::Pin, sync::Arc};
+
+use datafusion::{
+    arrow::{array::RecordBatch, datatypes::SchemaRef as ArrowSchemaRef},
+    execution::{SendableRecordBatchStream, TaskContext},
+    physical_expr::EquivalenceProperties,
+    physical_plan::{
+        stream::RecordBatchStreamAdapter, DisplayAs, ExecutionMode, ExecutionPlan, Partitioning,
+        PlanProperties,
+    },
+};
+use futures::{Stream, TryStreamExt};
+use iceberg::table::Table;
+
+use crate::to_datafusion_error;
+
+#[derive(Debug)]
+pub(crate) struct IcebergTableScan {
+    table: Table,
+    schema: ArrowSchemaRef,
+    plan_properties: PlanProperties,
+}
+
+impl IcebergTableScan {
+    pub(crate) fn new(table: Table, schema: ArrowSchemaRef) -> Self {
+        let plan_properties = Self::compute_properties(schema.clone());
+
+        Self {
+            table,
+            schema,
+            plan_properties,
+        }
+    }
+
+    fn compute_properties(schema: ArrowSchemaRef) -> PlanProperties {
+        // TODO:
+        // This is more or less a placeholder, to be replaced
+        // once we support output-partitioning
+        PlanProperties::new(
+            EquivalenceProperties::new(schema),
+            Partitioning::UnknownPartitioning(1),
+            ExecutionMode::Bounded,
+        )
+    }
+}
+
+impl ExecutionPlan for IcebergTableScan {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        _children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> datafusion::error::Result<Arc<dyn ExecutionPlan>> {
+        Ok(self)
+    }
+
+    fn properties(&self) -> &PlanProperties {
+        &self.plan_properties
+    }
+
+    fn execute(
+        &self,
+        _partition: usize,
+        _context: Arc<TaskContext>,
+    ) -> datafusion::error::Result<SendableRecordBatchStream> {
+        let fut = get_batch_stream(self.table.clone());
+        let stream = futures::stream::once(fut).try_flatten();
+
+        Ok(Box::pin(RecordBatchStreamAdapter::new(
+            self.schema.clone(),
+            stream,
+        )))
+    }
+}
+
+impl DisplayAs for IcebergTableScan {
+    fn fmt_as(
+        &self,
+        _t: datafusion::physical_plan::DisplayFormatType,
+        f: &mut std::fmt::Formatter,
+    ) -> std::fmt::Result {
+        write!(f, "IcebergTableScan")
+    }
+}
+
+async fn get_batch_stream(
+    table: Table,
+) -> datafusion::error::Result<
+    Pin<Box<dyn Stream<Item = datafusion::error::Result<RecordBatch>> + Send>>,
+> {
+    let table_scan = table.scan().build().map_err(to_datafusion_error)?;
+
+    let stream = table_scan
+        .to_arrow()
+        .await
+        .map_err(to_datafusion_error)?
+        .map_err(to_datafusion_error);
```

Comment:
@liurenjie1024 @ZENOTME @viirya 
I had no idea how to solve this better; but it still looks weird and redundant.

Why do I need to do this? First `map_err` unwraps the result from `to_arrow()` and the second `map_err` converts the outer result (wrapping the boxed stream) into a `datafusion::error:Result` which is required by `futures::stream::once(fut).try_flatten()`.

Reply:
I think it's ok for now. There is [flatten](https://doc.rust-lang.org/std/result/enum.Result.html#method.flatten) api to simplify it, but it's not stable. We can refine it in future.